File: casper/src/test/scala/coop/rchain/casper/batch2/LimitedParentDepthSpec.scala
package coop.rchain.casper.batch2

import cats.instances.list._
import cats.syntax.traverse._
import coop.rchain.casper.helper.TestNode
import coop.rchain.casper.util.ConstructDeploy.basicDeployData
import coop.rchain.casper.util.GenesisBuilder.buildGenesis
import coop.rchain.p2p.EffectsTestInstances.LogicalTime
import monix.eval.Task
import monix.execution.Scheduler
import org.scalatest.{FlatSpec, Matchers}

class LimitedParentDepthSpec extends FlatSpec with Matchers {
  implicit val scheduler = Scheduler.fixedPool("limited-parent-depth-scheduler", 2)
  implicit val timeEff   = new LogicalTime[Task]

  val genesisContext = buildGenesis()

  "Estimator" should "obey present parent depth limitation" in {
    val t = TestNode.networkEff(genesisContext, networkSize = 2, maxParentDepth = Some(2)).use {
      case nodes @ n1 +: n2 +: Seq() =>
        for {
          produceDeploys <- (0 until 6).toList
                             .traverse(i => basicDeployData[Task](i, shardId = "root"))

          b1 <- n1.propagateBlock(produceDeploys(0))()
          b2 <- n2.propagateBlock(produceDeploys(1))(nodes: _*)
          b4 <- n2.propagateBlock(produceDeploys(2))(nodes: _*)
          b4 <- n2.propagateBlock(produceDeploys(3))(nodes: _*)
          b5 <- n2.propagateBlock(produceDeploys(4))(nodes: _*)
          b6 <- n1.propagateBlock(produceDeploys(5))(nodes: _*)
        } yield b6.header.parentsHashList shouldBe List(b5.blockHash)
    }
    t.runSyncUnsafe()
  }

  it should "obey absent parent depth limitation" in {
    val t = TestNode.networkEff(genesisContext, networkSize = 2, maxParentDepth = None).use {
      case nodes @ n1 +: n2 +: Seq() =>
        for {
          produceDeploys <- (0 until 6).toList
                             .traverse(i => basicDeployData[Task](i, shardId = "root"))

          b1 <- n1.propagateBlock(produceDeploys(0))()
          b2 <- n2.propagateBlock(produceDeploys(1))(nodes: _*)
          b4 <- n2.propagateBlock(produceDeploys(2))(nodes: _*)
          b4 <- n2.propagateBlock(produceDeploys(3))(nodes: _*)
          b5 <- n2.propagateBlock(produceDeploys(4))(nodes: _*)
          b6 <- n1.propagateBlock(produceDeploys(5))(nodes: _*)
        } yield {
          // With multi-parent merging, all validators' latest blocks are included as parents.
          // Genesis has 4 validators, only 2 nodes create blocks, so validators 2 and 3 still
          // have genesis as their latest message. This results in 3 unique parents:
          // - b1 (from validator 0)
          // - b5 (from validator 1)
          // - genesis (from validators 2 and 3 who haven't created blocks)
          b6.header.parentsHashList should have size 3
          b6.header.parentsHashList should contain(b1.blockHash)
          b6.header.parentsHashList should contain(b5.blockHash)
        }
    }
    t.runSyncUnsafe()
  }
}

File: casper/tests/batch2/limited_parent_depth_spec.rs
// See casper/src/test/scala/coop/rchain/casper/batch2/LimitedParentDepthSpec.scala

use crate::helper::test_node::TestNode;
use crate::util::genesis_builder::GenesisBuilder;
use casper::rust::errors::CasperError;
use casper::rust::util::construct_deploy;
use crypto::rust::signatures::signed::Signed;
use models::rust::casper::protocol::casper_message::DeployData;

struct TestContext {
    genesis: crate::util::genesis_builder::GenesisContext,
    produce_deploys: Vec<Signed<DeployData>>,
}

impl TestContext {
    async fn new() -> Self {
        let genesis = GenesisBuilder::new()
            .build_genesis_with_parameters(None)
            .await
            .unwrap();

        let mut produce_deploys = Vec::new();
        for i in 0..6 {
            tokio::time::sleep(tokio::time::Duration::from_millis(1)).await;
            let deploy = construct_deploy::basic_deploy_data(
                i,
                None,
                Some(genesis.genesis_block.shard_id.clone()),
            )
            .unwrap();
            produce_deploys.push(deploy);
        }

        Self {
            genesis,
            produce_deploys,
        }
    }

    async fn create_network(
        &self,
        max_parent_depth: Option<i32>,
    ) -> Result<Vec<TestNode>, CasperError> {
        TestNode::create_network(self.genesis.clone(), 2, None, None, max_parent_depth, None).await
    }
}

#[tokio::test]
async fn estimator_should_obey_present_parent_depth_limitation() {
    let ctx = TestContext::new().await;
    let mut nodes = ctx.create_network(Some(2)).await.unwrap();

    let _b1 = nodes[0]
        .add_block_from_deploys(&[ctx.produce_deploys[0].clone()])
        .await
        .unwrap();

    let _b2 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[1].clone()])
        .await
        .unwrap();

    let _b3 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[2].clone()])
        .await
        .unwrap();

    let _b4 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[3].clone()])
        .await
        .unwrap();

    let b5 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[4].clone()])
        .await
        .unwrap();

    let b6 = TestNode::propagate_block_at_index(&mut nodes, 0, &[ctx.produce_deploys[5].clone()])
        .await
        .unwrap();

    assert_eq!(
        b6.header.parents_hash_list,
        vec![b5.block_hash.clone()],
        "Expected b6 to have only b5 as parent due to maxParentDepth=2"
    );
}

#[tokio::test]
async fn estimator_should_obey_absent_parent_depth_limitation() {
    let ctx = TestContext::new().await;
    let mut nodes = ctx.create_network(None).await.unwrap();

    let b1 = nodes[0]
        .add_block_from_deploys(&[ctx.produce_deploys[0].clone()])
        .await
        .unwrap();

    let _b2 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[1].clone()])
        .await
        .unwrap();

    let _b3 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[2].clone()])
        .await
        .unwrap();

    let _b4 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[3].clone()])
        .await
        .unwrap();

    let b5 = TestNode::propagate_block_at_index(&mut nodes, 1, &[ctx.produce_deploys[4].clone()])
        .await
        .unwrap();

    let b6 = TestNode::propagate_block_at_index(&mut nodes, 0, &[ctx.produce_deploys[5].clone()])
        .await
        .unwrap();

    // With multi-parent merging, all validators' latest blocks are included as parents.
    // Genesis has 4 validators, only 2 nodes create blocks, so validators 2 and 3 still
    // have genesis as their latest message. This results in 3 unique parents:
    // - b1 (from validator 0)
    // - b5 (from validator 1)
    // - genesis (from validators 2 and 3 who haven't created blocks)
    assert_eq!(
        b6.header.parents_hash_list.len(),
        3,
        "Expected b6 to have exactly 3 parents (b1, b5, genesis)"
    );
    assert!(
        b6.header.parents_hash_list.contains(&b1.block_hash),
        "Expected b6 to have b1 as a parent"
    );
    assert!(
        b6.header.parents_hash_list.contains(&b5.block_hash),
        "Expected b6 to have b5 as a parent"
    );
}
